system/security/plugins/jwtSecurity/jwtEndpoint.cpp

```diff
@@ -0,0 +1,230 @@
+#include <string.h>
+
+#include "jlog.hpp"
+
+#include "nlohmann/json.hpp"
+
+#include "jwtEndpoint.hpp"
+
+struct MemoryStruct
+{
+    char*   memory;
+    size_t  size;
+    size_t  capacity;
+};
+
+static void initMemoryStructCapacity(MemoryStruct& mem, size_t initCapacity)
+{
+    mem.memory = static_cast<char*>(malloc(initCapacity));
+    mem.size = 0;
+    mem.capacity = initCapacity;
+}
+
+static void freeMemoryStruct(MemoryStruct& mem)
+{
+    if (mem.memory)
+    {
+        free(mem.memory);
+        mem.memory = nullptr;
+    }
+    mem.size = 0;
+    mem.capacity = 0;
+}
+
+static size_t captureIncomingCURLReply(void* contents, size_t size, size_t nmemb, void* userp)
+{
+    size_t          incomingDataSize = size * nmemb;
+    MemoryStruct*   mem = static_cast<struct MemoryStruct*>(userp);
+    size_t          MAX_BUFFER_SIZE = 4194304; // 2^22
+
+    if ((mem->size + incomingDataSize) < MAX_BUFFER_SIZE)
+    {
+        if ((mem->size + incomingDataSize) > mem->capacity)
+        {
+            size_t  newCapacity = mem->capacity * 2;
+
+            // Keep doubling capacity until it is greater than what we need
+            while ((mem->size + incomingDataSize) > newCapacity)
+                newCapacity *= 2;
+
+            mem->memory = static_cast<char*>(realloc(mem->memory, newCapacity));
+            mem->capacity = newCapacity;
+        }
+
+        memcpy(&(mem->memory[mem->size]), contents, incomingDataSize);
+        mem->size += incomingDataSize;
+    }
+    else
+    {
+        // Signals an error to libcurl
+        incomingDataSize = 0;
+
+        PROGLOG("CJwtSecurityManager::captureIncomingCURLReply() exceeded buffer size %ld", MAX_BUFFER_SIZE);
+    }
+
+    return incomingDataSize;
+}
+
+static std::string hashString(const std::string& s)
+{
+    SHA256_CTX  context;
+    char        hashedValue[SHA256_DIGEST_LENGTH + 1];
+
+    memset(hashedValue, 0, sizeof(hashedValue));
+
+    if (!SHA256_Init(&context))
+        throw MakeStringException(-1, "CJwtSecurityManager: OpenSSL ERROR calling SHA256_Init while hashing user password");
+
+    if (!SHA256_Update(&context, (unsigned char*)s.c_str(), s.size()))
+        throw MakeStringException(-1, "CJwtSecurityManager: OpenSSL ERROR calling SHA256_Update while hashing user password");
+
+    if (!SHA256_Final((unsigned char*)hashedValue, &context))
+        throw MakeStringException(-1, "CJwtSecurityManager: OpenSSL ERROR calling SHA256_Final while hashing user password");
+
+    return hashedValue;
+}
+
+static std::string hashUserPW(const std::string& pw, const std::string& nonce)
+{
+    std::string     firstHash(hashString(pw));
+    std::string     secondHash(hashString(firstHash + nonce));
+
+    return secondHash;
+}
+
+static std::string tokenFromEndpoint(const std::string& jwtEndPoint, bool allowSelfSignedCert, const std::string& credentialsStr)
+{
+    std::string     apiResponse;
+
+    // Call JWT endpoint
+    CURL*   curlHandle = curl_easy_init();
```

Comment:
How expensive is this call?  Should it be cached in a thread local variable?

Reply:
Initializing libcurl's handle is not expensive, though some memory (for state) is set aside and structs initialized.  This is called very infrequently, only during initial user login and during token refreshes.